raft/raft.go

```diff
@@ -1397,24 +1437,48 @@ func (r *raft) promotable() bool {
 	return pr != nil && !pr.IsLearner
 }
 
-func (r *raft) applyConfChange(cc pb.ConfChange) pb.ConfState {
-	cfg, prs, err := confchange.Changer{
-		Tracker:   r.prs,
-		LastIndex: r.raftLog.lastIndex(),
-	}.Simple(pb.ConfChangeSingle{
-		Type:   cc.Type,
-		NodeID: cc.NodeID,
-	})
+func (r *raft) applyConfChange(cc pb.ConfChangeV2) pb.ConfState {
+	cfg, prs, err := func() (tracker.Config, tracker.ProgressMap, error) {
+		changer := confchange.Changer{
+			Tracker:   r.prs,
+			LastIndex: r.raftLog.lastIndex(),
+		}
+		if cc.LeaveJoint() {
+			return changer.LeaveJoint()
+		} else if cc.EnterJoint() {
+			var autoLeave bool
+			switch cc.Transition {
```

Comment:
The fact that `cc.EnterJoint` is a function of `cc.Transition` and then we immediately do another switch on `cc.Transition` smells a little bad. Maybe `cc.EnterJoint` should return the autoleave value to centralize knowledge of the transition types?

Reply:
I don't want to return `(bool, bool)` but it looks like returning a `*bool` works, see if you prefer that.

Reply:
Ugh, I like `(bool, bool)` better. With that version I have to check the docs to make sure I have the right order, but after that the code is clear. With `*bool` I have to think more about what's going on.

Reply:
I can't say that I like this more, but I made the change.